File: AUGUST-2025/06-08-2025/p3479.ts
function numOfUnplacedFruits(fruits: number[], baskets: number[]): number {
    const n = baskets.length;
    const blockSize = Math.floor(Math.sqrt(n)) + 1;
    const numBlocks = Math.ceil(n / blockSize);
    const maxV = Array(numBlocks).fill(0);

    // Precompute max in each block
    for (let i = 0; i < n; ++i) {
        const block = Math.floor(i / blockSize);
        maxV[block] = Math.max(maxV[block], baskets[i]);
    }

    let unplaced = 0;

    for (const fruit of fruits) {
        let placed = false;
        for (let b = 0; b < numBlocks; ++b) {
            if (maxV[b] < fruit) continue;

            const start = b * blockSize;
            const end = Math.min(n, start + blockSize);

            for (let i = start; i < end; ++i) {
                if (baskets[i] >= fruit) {
                    baskets[i] = 0;

                    // Recompute block max
                    maxV[b] = 0;
                    for (let j = start; j < end; ++j) {
                        maxV[b] = Math.max(maxV[b], baskets[j]);
                    }

                    placed = true;
                    break;
                }
            }

            if (placed) break;
        }

        if (!placed) unplaced++;
    }

    return unplaced;
}

File: AUGUST-2025/06-08-2025/p3479.cpp
class Solution {
public:
    void build(int i, int l, int r, vector<int>& segment,
               vector<int>& baskets) {
        if (l == r) {
            segment[i] = baskets[l];
            return;
        }

        int md = l + (r - l) / 2;
        build(2 * i + 1, l, md, segment, baskets);
        build(2 * i + 2, md + 1, r, segment, baskets);

        segment[i] = max(segment[2 * i + 1], segment[2 * i + 2]);
    }

    bool segmentQuery(int i, int l, int r, vector<int>& segment, int fruit) {

        if (segment[i] < fruit)
            return false;

        if (l == r) {
            segment[i] = -1;
            return true;
        }
        int md = l + (r - l) / 2;
        bool placed = false;
        if (segment[2 * i + 1] >= fruit) {
            placed = segmentQuery(2 * i + 1, l, md, segment, fruit);
        } else
            placed = segmentQuery(2 * i + 2, md + 1, r, segment, fruit);

        segment[i] = max(segment[2 * i + 1], segment[2 * i + 2]);
        return placed;
    }

    int numOfUnplacedFruits(vector<int>& fruits, vector<int>& baskets) {
        int n = baskets.size();
        vector<int> segment(4 * n, -1);

        build(0, 0, n - 1, segment, baskets);

        int unplaced = 0;

        for (int& fr : fruits) {
            if (segmentQuery(0, 0, n - 1, segment, fr) == false)
                unplaced++;
        }
        return unplaced;
    }
};